packages/common/src/tools/queue/TaskQueue.ts

```diff
@@ -0,0 +1,113 @@
+import assert from 'assert'
+
+import { Logger } from '../Logger'
+import { Retries } from './Retries'
+import { Job, ShouldRetry } from './types'
+
+interface TaskQueueOpts<T> {
+  workers?: number
+  shouldRetry?: ShouldRetry<T>
+}
+
+export class TaskQueue<T> {
+  private queue: Job<T>[] = []
+  private busyWorkers = 0
+  private workers: number
+  private shouldRetry: ShouldRetry<T>
+
+  constructor(
+    private executeTask: (task: T) => Promise<void>,
+    private logger: Logger,
+    opts?: TaskQueueOpts<T>,
+  ) {
+    this.workers = opts?.workers ?? 1
+    assert(
+      this.workers > 0 && Number.isInteger(this.workers),
+      'workers needs to be a positive integer',
+    )
+    this.shouldRetry =
+      opts?.shouldRetry ??
+      Retries.exponentialBackOff(100, { maxAttempts: 10, maxDistance: 10_000 })
+  }
+
+  private get isEmpty() {
+    return this.queue.length === 0 && this.busyWorkers === 0
+  }
+
+  private get allWorkersBusy() {
+    return this.busyWorkers === this.workers
+  }
+
+  addIfEmpty(task: T) {
+    if (!this.isEmpty) {
+      return
+    }
+    this.addToBack(task)
+  }
+
+  addToFront(task: T) {
+    this.queue.unshift({ task, attempts: 0, executeAfter: Date.now() })
+    setTimeout(() => this.execute())
+  }
+
+  addToBack(task: T) {
+    this.queue.push({ task, attempts: 0, executeAfter: Date.now() })
+    setTimeout(() => this.execute())
+  }
+
+  private execute() {
+    this.executeUnchecked().catch((e) => {
+      this.logger.error(e)
+    })
+  }
+
+  private shouldExecute(job: Job<T>): boolean {
+    return Date.now() >= job.executeAfter
+  }
+
+  private handleFailure(job: Job<T>) {
+    job.attempts++
+    const result = this.shouldRetry(job)
+    if (!result.retry) {
+      this.logger.debug('Task should not be retried - removing from queue', {
+        task: String(job.task),
+        attempts: job.attempts,
+      })
+      return
+    }
+    job.executeAfter = result.executeAfter ?? Date.now()
+    this.logger.debug('Retrying task', {
+      task: String(job.task),
```

Comment:
I would prefer if the tasks had an explicit textual description. This solution will most likely print `[Object object]` or some other nonsense

Reply:
Unfortunately there is no guarantee that `JSON.stringify` doesn't throw. For example it will throw if the task includes a bignumber. This is why we have the `json` type. If you'd like to have a default description you could have the queue keep track of an incrementing id and give a default description of `Task #${++id}`. And the signature would be `addToBack(task: T, description?: string)`